Two's complement representation for decimals

martint opened the pull request.

martint committed:
Remove unnecessary branch

Both branches call the same method, so one of them is not needed.

martint committed:
Use 2's complement representation for decimals

Also, use a dedicated class to represent long decimals instead of Slice.

martint commented:
Some preliminary benchmark results for decimal addition:

```
before

Benchmark                                             (expression)   Mode  Cnt  Score   Error   Units
BenchmarkDecimalOperators.additionBenchmark                l1 + l2  thrpt   15  1.055 ± 0.002  ops/ms
BenchmarkDecimalOperators.additionBenchmark      l1 + l2 + l3 + l4  thrpt   15  0.429 ± 0.006  ops/ms
BenchmarkDecimalOperators.additionBenchmark      s2 + l3 + l1 + s4  thrpt   15  0.433 ± 0.007  ops/ms
BenchmarkDecimalOperators.additionBenchmark              lz1 + lz2  thrpt   15  1.121 ± 0.008  ops/ms
BenchmarkDecimalOperators.additionBenchmark  lz1 + lz2 + lz3 + lz4  thrpt   15  0.469 ± 0.007  ops/ms
BenchmarkDecimalOperators.additionBenchmark    s2 + lz3 + lz1 + s4  thrpt   15  0.311 ± 0.022  ops/ms


after

Benchmark                                             (expression)   Mode  Cnt  Score   Error   Units
BenchmarkDecimalOperators.additionBenchmark                l1 + l2  thrpt   15  2.270 ± 0.011  ops/ms
BenchmarkDecimalOperators.additionBenchmark      l1 + l2 + l3 + l4  thrpt   15  1.103 ± 0.003  ops/ms
BenchmarkDecimalOperators.additionBenchmark      s2 + l3 + l1 + s4  thrpt   15  1.015 ± 0.014  ops/ms
BenchmarkDecimalOperators.additionBenchmark              lz1 + lz2  thrpt   15  3.599 ± 0.011  ops/ms
BenchmarkDecimalOperators.additionBenchmark  lz1 + lz2 + lz3 + lz4  thrpt   15  2.571 ± 0.216  ops/ms
BenchmarkDecimalOperators.additionBenchmark    s2 + lz3 + lz1 + s4  thrpt   15  1.068 ± 0.017  ops/ms
```

findepi commented:
If this changes long decimal representation, how do we address connectors compatibility?

martint commented:
> If this changes long decimal representation, how do we address connectors compatibility?

I don't think there's a good way around that. We'll probably have to make this a backward compatibility breaking change. It's unfortunate that we chose the current representation when we first introduced decimals, and that we didn't have the facilities to support arbitrary object types other than Slice back then (which seems to introduce non-trivial performance issues).

On the positive side, since we're changing the stack type, any connector that doesn't support the new representation will just fail loudly instead of producing incorrect results.

There may be ways to do it in a backward compatible way, though -- I haven't spent too much time thinking about it yet, though. If you have any thoughts on how we might do it, please comment.

kokosing commented:
How about exposing method like `io.trino.spi.connector.Connector#isUsingLegacyDecimalRepresentation`, if so during planning we could add a projection that would call a scalar function that would translate decimal between representations. That way we could migrate one connector after the other.

martint commented:
Not sure how that would work. The type can only support one “stack” and block type

findepi commented:
@kokosing actually i am in favor of backwards incompatible approach.
desire to maintain backwards compatibility is what was preventing us from fixing timestamp semantics for so long. 
We were able to fix (most of) #37 not because we found smarter ways to achieve the goals, but rather: we redefined the goals.
OTOH, i would be surprised if "others" were OK with backwards incompatible approach as well. I remember the intense drama related to #9765  which was not a real problem and in a new API implemented by no-one, especially compared to big-bang approach proposed here, for type representation used by nearly every connector.

kokosing commented:
> Not sure how that would work. The type can only support one “stack” and block type

I am sorry I haven't read the code yet. I was under impression that "stack" type didn't change but only encoding.

martint commented:
Update benchmarks:

```
*** Before

Benchmark                                                                                                            (expression)  (precision)   Mode  Cnt  Score    Error   Units
BenchmarkDecimalOperators.additionBenchmark                                                                               s1 + s2          N/A  thrpt   15  6.304 ±  0.038  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                     s1 + s2 + s3 + s4          N/A  thrpt   15  1.136 ±  0.004  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                               l1 + l2          N/A  thrpt   15  1.558 ±  0.021  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                     l1 + l2 + l3 + l4          N/A  thrpt   15  0.399 ±  0.014  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                     s2 + l3 + l1 + s4          N/A  thrpt   15  0.331 ±  0.003  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                             lz1 + lz2          N/A  thrpt   15  1.602 ±  0.007  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                 lz1 + lz2 + lz3 + lz4          N/A  thrpt   15  0.466 ±  0.001  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                   s2 + lz3 + lz1 + s4          N/A  thrpt   15  0.317 ±  0.021  ops/ms
BenchmarkDecimalOperators.castDecimalToDoubleBenchmark                                                                        N/A           15  thrpt   15  9.778 ±  0.040  ops/ms
BenchmarkDecimalOperators.castDecimalToDoubleBenchmark                                                                        N/A           35  thrpt   15  0.035 ±  0.001  ops/ms
BenchmarkDecimalOperators.castDecimalToVarcharBenchmark                                                                       N/A           15  thrpt   15  0.256 ±  0.003  ops/ms
BenchmarkDecimalOperators.castDecimalToVarcharBenchmark                                                                       N/A           35  thrpt   15  0.057 ±  0.001  ops/ms
BenchmarkDecimalOperators.castDoubleToDecimalBenchmark                                                                        N/A           10  thrpt   15  0.103 ±  0.008  ops/ms
BenchmarkDecimalOperators.castDoubleToDecimalBenchmark                                                                        N/A           35  thrpt   15  0.047 ±  0.005  ops/ms
BenchmarkDecimalOperators.castDoubleToDecimalBenchmark                                                                        N/A       BIGINT  thrpt   15  7.244 ±  0.014  ops/ms
BenchmarkDecimalOperators.decimalToShortDecimalCastBenchmark                                       cast(l_38_30 as decimal(8, 0))          N/A  thrpt   15  0.798 ±  0.004  ops/ms
BenchmarkDecimalOperators.decimalToShortDecimalCastBenchmark                                       cast(l_26_18 as decimal(8, 0))          N/A  thrpt   15  1.890 ±  0.010  ops/ms
BenchmarkDecimalOperators.decimalToShortDecimalCastBenchmark                                       cast(l_20_12 as decimal(8, 0))          N/A  thrpt   15  1.133 ±  0.016  ops/ms
BenchmarkDecimalOperators.decimalToShortDecimalCastBenchmark                                        cast(l_20_8 as decimal(8, 0))          N/A  thrpt   15  1.146 ±  0.059  ops/ms
BenchmarkDecimalOperators.decimalToShortDecimalCastBenchmark                                        cast(s_17_9 as decimal(8, 0))          N/A  thrpt   15  8.692 ±  0.056  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               s1 / s2          N/A  thrpt   15  6.152 ±  0.048  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                     s1 / s2 / s2 / s2          N/A  thrpt   15  0.888 ±  0.007  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               s1 / s3          N/A  thrpt   15  0.275 ±  0.002  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               s2 / l1          N/A  thrpt   15  0.302 ±  0.003  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               l1 / s2          N/A  thrpt   15  0.673 ±  0.006  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               s3 / l1          N/A  thrpt   15  0.346 ±  0.005  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               l2 / l3          N/A  thrpt   15  0.180 ±  0.004  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                     l2 / l4 / l4 / l4          N/A  thrpt   15  0.108 ±  0.001  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                     l2 / s4 / s4 / s4          N/A  thrpt   15  0.117 ±  0.002  ops/ms
BenchmarkDecimalOperators.inequalityBenchmark                                                                             s1 < s2          N/A  thrpt   15  6.730 ±  0.076  ops/ms
BenchmarkDecimalOperators.inequalityBenchmark                 s1 < s2 AND s1 < s3 AND s1 < s4 AND s2 < s3 AND s2 < s4 AND s3 < s4          N/A  thrpt   15  4.354 ±  0.034  ops/ms
BenchmarkDecimalOperators.inequalityBenchmark                                                                             l1 < l2          N/A  thrpt   15  6.056 ±  0.059  ops/ms
BenchmarkDecimalOperators.inequalityBenchmark                 l1 < l2 AND l1 < l3 AND l1 < l4 AND l2 < l3 AND l2 < l4 AND l3 < l4          N/A  thrpt   15  2.753 ±  0.052  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                                 s1 % s2          N/A  thrpt   15  1.071 ±  0.003  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                       s1 % s2 % s2 % s2          N/A  thrpt   15  0.133 ±  0.003  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                                 s2 % l2          N/A  thrpt   15  0.764 ±  0.005  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                                 l3 % s3          N/A  thrpt   15  0.719 ±  0.005  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                                 s4 % l3          N/A  thrpt   15  0.852 ±  0.009  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                                 l2 % l3          N/A  thrpt   15  0.403 ±  0.003  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                       l2 % l3 % l4 % l1          N/A  thrpt   15  0.173 ±  0.001  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                               s1 * s2          N/A  thrpt   15  7.809 ±  0.024  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                     s1 * s2 * s5 * s6          N/A  thrpt   15  2.557 ±  0.021  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                               s3 * s4          N/A  thrpt   15  2.048 ±  0.036  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                               l2 * s2          N/A  thrpt   15  1.716 ±  0.010  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                     l2 * s2 * s5 * s6          N/A  thrpt   15  1.042 ±  0.003  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                               s1 * l2          N/A  thrpt   15  1.609 ±  0.017  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                               l1 * l2          N/A  thrpt   15  1.634 ±  0.016  ops/ms

*** Int128

Benchmark                                                                                                            (expression)  (precision)   Mode  Cnt  Score    Error   Units
BenchmarkDecimalOperators.additionBenchmark                                                                               s1 + s2          N/A  thrpt   15  6.160 ±  0.108  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                     s1 + s2 + s3 + s4          N/A  thrpt   15  5.067 ±  0.038  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                               l1 + l2          N/A  thrpt   15  2.154 ±  0.026  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                     l1 + l2 + l3 + l4          N/A  thrpt   15  0.510 ±  0.003  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                     s2 + l3 + l1 + s4          N/A  thrpt   15  1.012 ±  0.002  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                             lz1 + lz2          N/A  thrpt   15  4.932 ±  0.013  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                 lz1 + lz2 + lz3 + lz4          N/A  thrpt   15  2.884 ±  0.007  ops/ms
BenchmarkDecimalOperators.additionBenchmark                                                                   s2 + lz3 + lz1 + s4          N/A  thrpt   15  1.095 ±  0.006  ops/ms
BenchmarkDecimalOperators.castDecimalToDoubleBenchmark                                                                        N/A           15  thrpt   15  9.883 ±  0.040  ops/ms
BenchmarkDecimalOperators.castDecimalToDoubleBenchmark                                                                        N/A           35  thrpt   15  0.060 ±  0.001  ops/ms
BenchmarkDecimalOperators.castDecimalToVarcharBenchmark                                                                       N/A           15  thrpt   15  0.266 ±  0.003  ops/ms
BenchmarkDecimalOperators.castDecimalToVarcharBenchmark                                                                       N/A           35  thrpt   15  0.102 ±  0.001  ops/ms
BenchmarkDecimalOperators.castDoubleToDecimalBenchmark                                                                        N/A           10  thrpt   15  0.115 ±  0.010  ops/ms
BenchmarkDecimalOperators.castDoubleToDecimalBenchmark                                                                        N/A           35  thrpt   15  0.056 ±  0.003  ops/ms
BenchmarkDecimalOperators.castDoubleToDecimalBenchmark                                                                        N/A       BIGINT  thrpt   15  7.106 ±  0.033  ops/ms
BenchmarkDecimalOperators.decimalToShortDecimalCastBenchmark                                       cast(l_38_30 as decimal(8, 0))          N/A  thrpt   15  2.510 ±  0.006  ops/ms
BenchmarkDecimalOperators.decimalToShortDecimalCastBenchmark                                       cast(l_26_18 as decimal(8, 0))          N/A  thrpt   15  3.616 ±  0.005  ops/ms
BenchmarkDecimalOperators.decimalToShortDecimalCastBenchmark                                       cast(l_20_12 as decimal(8, 0))          N/A  thrpt   15  5.530 ±  0.026  ops/ms
BenchmarkDecimalOperators.decimalToShortDecimalCastBenchmark                                        cast(l_20_8 as decimal(8, 0))          N/A  thrpt   15  5.389 ±  0.023  ops/ms
BenchmarkDecimalOperators.decimalToShortDecimalCastBenchmark                                        cast(s_17_9 as decimal(8, 0))          N/A  thrpt   15  8.677 ±  0.077  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               s1 / s2          N/A  thrpt   15  6.093 ±  0.020  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                     s1 / s2 / s2 / s2          N/A  thrpt   15  2.473 ±  0.005  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               s1 / s3          N/A  thrpt   15  0.646 ±  0.007  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               s2 / l1          N/A  thrpt   15  1.104 ±  0.005  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               l1 / s2          N/A  thrpt   15  0.832 ±  0.003  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               s3 / l1          N/A  thrpt   15  0.649 ±  0.005  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                               l2 / l3          N/A  thrpt   15  0.443 ±  0.002  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                     l2 / l4 / l4 / l4          N/A  thrpt   15  0.264 ±  0.001  ops/ms
BenchmarkDecimalOperators.divisionBenchmark                                                                     l2 / s4 / s4 / s4          N/A  thrpt   15  0.322 ±  0.001  ops/ms
BenchmarkDecimalOperators.inequalityBenchmark                                                                             s1 < s2          N/A  thrpt   15  6.674 ±  0.028  ops/ms
BenchmarkDecimalOperators.inequalityBenchmark                 s1 < s2 AND s1 < s3 AND s1 < s4 AND s2 < s3 AND s2 < s4 AND s3 < s4          N/A  thrpt   15  4.346 ±  0.011  ops/ms
BenchmarkDecimalOperators.inequalityBenchmark                                                                             l1 < l2          N/A  thrpt   15  6.055 ±  0.055  ops/ms
BenchmarkDecimalOperators.inequalityBenchmark                 l1 < l2 AND l1 < l3 AND l1 < l4 AND l2 < l3 AND l2 < l4 AND l3 < l4          N/A  thrpt   15  2.499 ±  0.011  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                                 s1 % s2          N/A  thrpt   15  1.383 ±  0.006  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                       s1 % s2 % s2 % s2          N/A  thrpt   15  0.682 ±  0.002  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                                 s2 % l2          N/A  thrpt   15  1.278 ±  0.007  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                                 l3 % s3          N/A  thrpt   15  0.407 ±  0.018  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                                 s4 % l3          N/A  thrpt   15  0.473 ±  0.005  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                                 l2 % l3          N/A  thrpt   15  0.735 ±  0.003  ops/ms
BenchmarkDecimalOperators.moduloBenchmark                                                                       l2 % l3 % l4 % l1          N/A  thrpt   15  0.260 ±  0.001  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                               s1 * s2          N/A  thrpt   15  7.696 ±  0.051  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                     s1 * s2 * s5 * s6          N/A  thrpt   15  5.113 ±  0.022  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                               s3 * s4          N/A  thrpt   15  1.856 ±  0.018  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                               l2 * s2          N/A  thrpt   15  1.939 ±  0.043  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                     l2 * s2 * s5 * s6          N/A  thrpt   15  1.041 ±  0.008  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                               s1 * l2          N/A  thrpt   15  2.073 ±  0.023  ops/ms
BenchmarkDecimalOperators.multiplyBenchmark                                                                               l1 * l2          N/A  thrpt   15  1.995 ±  0.013  ops/ms
```

https://jmh.morethan.io/?sources=https://gist.githubusercontent.com/martint/23d897b83c4a54ad838085ee4466e0df/raw/2be351f0390777bc27ba0796648f425beab2f718/before.json,https://gist.githubusercontent.com/martint/23d897b83c4a54ad838085ee4466e0df/raw/2be351f0390777bc27ba0796648f425beab2f718/after.json

martint committed:
Improve decoding performance for long decimals

Avoid using intermediate BigIntegers when decoding long
decimals from Parquet and RCBinary

skrzypo987 commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+
+        boolean needsRounding;
+        if (shift < 64) {
+            needsRounding = roundUp && (low & 1L << (shift - 1)) != 0;
```
Put `1L << (shift - 1)` into parenthesis

skrzypo987 commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+        if (shift < 64) {
+            needsRounding = roundUp && (low & 1L << (shift - 1)) != 0;
+
+            low = (high << 1 << ~shift) | (low >>> shift);
```
Why double `<<` instead of `<< (1 - shift)`?

martint replied:
Actually, I think this is just equivalent to `-shift`:

`-shift = ~shift + 1` => `~shift = -shift - 1` => `(high << 1 << ~shift) == high << 1 - shift - 1 == high << -shift`.

However, I need to check if it's equivalent for some of the edge cases of shifts in java, which is really: `v >> s == v >> (s % 64)` (e.g., `-1 >> 64 == -1`, `16L >> 65 == 8`, etc.)

martint replied:
Indeed,

```
1. high << 1 << ~shift
```

and

```
2. high << 1 + (-shift - 1) 
    == high << 1 - shift - 1
    == high << -shift 
    == high << 64 - shift
```

are not quite equivalent.

When `shift == 0`

```
1. high << 1 << -1 == high << 1 << 64 - 1 == high << 1 << 63
```

But,

```
2. high << 64 - 0 == high << 64 == high << 0
```

skrzypo987 commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+        }
+
+        if (needsRounding) {
+            long tmpHigh = incrementHigh(high, low);
```
local variables seems redundant

martint replied:
They are not. They are there to avoid any inadvertent dependencies between the result of the first call and the second call.

skrzypo987 commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+            int remainder = divideUnsignedMultiPrecision(dividend, dividendLength, divisor[0]);
+            checkState(dividend[dividend.length - 1] == 0);
+            arraycopy(dividend, 0, quotient, 0, quotient.length);
+            fill(dividend, 0);
```
Same here

martint replied:
Preexisting

skrzypo987 commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+        if (divisorLength == 1) {
+            int remainder = divideUnsignedMultiPrecision(dividend, dividendLength, divisor[0]);
+            checkState(dividend[dividend.length - 1] == 0);
+            arraycopy(dividend, 0, quotient, 0, quotient.length);
```
If quotient length is constant you can make a static copy method

martint replied:
Preexisting

skrzypo987 commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+     * Divides mutableDividend / mutable divisor
+     * Places quotient in first argument and reminder in first argument
+     */
+    private static void divideUnsignedMultiPrecision(int[] dividend, int[] divisor, int[] quotient)
```
Has there been any changes related to division logic compared to the old UnscaledDecimal128Arithmetic?

martint replied:
No, it’s the same as before

skrzypo987 commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+        }
+        // wordShifts = shifts / 32
+        int wordShifts = shifts >>> 5;
+        // we don't wan't to loose any leading bits
```
typo - wan't

martint replied:
Preexisting

skrzypo987 reviewed (commented):
Reviewed Int128 and Int128Math so far.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+    private static final Int128[] POWERS_OF_FIVE = new Int128[54]; // 5^54 is the largest value < Int128.MAX_VALUE
+
+    private static final long ALL_BITS_SET_64 = 0xFFFFFFFFFFFFFFFFL;
+    private static final long INT_BASE = 1L << 32;
```
This is preexisting. I haven’t touched division yet, although I have some improvements I’ll put up after we merge this PR

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+        boolean negative = high < 0;
+
+        if (negative) {
+            long tmpHigh = negateHighExact(high, low);
```
No, that’s intentional. It’s a pattern I’ve been using to avoid creating intermediate arrays wherever possible. Sometimes the VM can elide those allocations，but often times it can’t, and it’s hard to predict.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+
+    private static void multiply(long leftHigh, long leftLow, long rightHigh, long rightLow, long[] result, int offset)
+    {
+        boolean leftNegative = leftHigh < 0;
```
It’s actually not *that* straightforward, especially the conditions to detect overflow. I have another change in the pipeline to improve this, but that will come later.

The goal of this PR is to make the initial transition to two’s complement. I’ve implemented some operations that are straightforward, but I’ve left others alone for now.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+
+    public static Int128 multiply(long left, long right)
+    {
+        boolean rightNegative = right < 0;
```
See reply above

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+            long divisor = longTenToNth(scaleFactor);
+            long newLow = low / divisor;
+            if (low % divisor >= (divisor >> 1)) {
+                newLow++;
```
That’s preexisting. I haven’t modified the logic in the division code yet — that will come in a separate PR

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+    }
+
+    /**
+     * Scale down the value for 10**tenScale (this := this / 5**tenScale). This
```
Preexisting

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+        long resultHigh = z1High + z2Low + z3Low;
+
+        if ((leftHigh != 0 && rightHigh != 0) ||
+                resultHigh < 0 || z1High < 0 || z2Low < 0 || z3Low < 0 ||
```
Preexisting. Also, that was intentional, since the terms in that line are somewhat related.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1331 @@
+     */
+    private static void scaleDownTen(long high, long low, int tenScale, long[] result, int offset, boolean roundUp)
+    {
+        if (high < 0) {
```
Just a sanity check, since this method won't work if the value is positive and will produce incorrect results.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128.java:
```diff
@@ -0,0 +1,185 @@
+    public static int compare(long leftHigh, long leftLow, long rightHigh, long rightLow)
+    {
+        int comparison = Long.compare(leftHigh, rightHigh);
+        if (comparison == 0) {
```
You mean like this?

```java
int high = Long.compare(leftHigh, rightHigh);
int low = Long.compareUnsigned(leftLow, rightLow);
return high == 0 ? low : high;
```

It's not clear that that's actually generates branchless code (need to check the assembly), and it has extra cos of computing low unconditionally if the VM is not smart enough to move the computation inside the `high == 0` check if it's not able to generate branchless code.

skrzypo987 replied:
Rather something like:
```
int highComparison = Long.compare(leftHigh, rightHigh);
int lowComparison = Long.compareUnsigned(leftLow, rightLow);
comparison = highComparison + ~(highComparison & 1) * lowComparison;
```
But I guess you can put it on the todo list and benchmark it later

martint replied:
> `highComparison & 1`

That may not work. Long.compare is not guaranteed to return -1, 0 or 1. According to its javadocs, it will return "a value less than 0 if x < y", "a value greater than 0 if x > y".

It would need to be adjusted to something like:

```java
int high = Long.compare(aHigh, bHigh);
int low = Long.compareUnsigned(aLow, bLow);

int mask = ~(high | -high) >> 31;
int comparison = high + (mask & low);
```

skrzypo987 commented on core/trino-spi/src/main/java/io/trino/spi/type/LongDecimalType.java:
```diff
@@ -118,59 +110,39 @@ public void appendTo(Block block, int position, BlockBuilder blockBuilder)
-                left.getLong(SIZE_OF_LONG),
-                right.getLong(0),
-                right.getLong(SIZE_OF_LONG));
+        return left.getHigh() == right.getHigh() && left.getLow() == right.getLow();
```
You can use `Int128::equals`

skrzypo987 reviewed (commented):
Went through all of it. Looks good.

martint commented on core/trino-main/src/main/java/io/trino/type/DecimalCasts.java:
```diff
@@ -166,9 +162,9 @@ public static boolean shortDecimalToBoolean(long decimal, long precision, long s
+    public static boolean longDecimalToBoolean(Int128 decimal, long precision, long scale, BigInteger tenToScale)
     {
-        return !decodeUnscaledValue(decimal).equals(ZERO);
+        return !decimal.toBigInteger().equals(ZERO);
```
No particular reason. Some leftover from an automated refactoring.

dain commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1341 @@
+        return (int) qhat;
+    }
+
+    private static long divideUnsignedLong(long dividend, int divisor)
```
Can this be `Long.divideUnsigned()`

martint replied:
Yes, but it's pre-existent. I'll change it once I get the new division code in.

dain commented on core/trino-spi/src/test/java/io/trino/spi/type/TestInt128Math.java:
```diff
@@ -0,0 +1,818 @@
+
+    private static void assertMultiply(long a, long b, long result)
+    {
+        assertMultiply(a, b, BigInteger.valueOf(result));
```
If the implementation is supposed to match `BigInteger` you could use that to create the result value.

martint replied:
It's preexistent. I just renamed the class.

dain reviewed (approved):
Mostly minor comments

martint commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java:
```diff
@@ -142,20 +140,16 @@ public static void inputShortDecimal(LongDecimalWithOverflowState state, Block b
                 decimal,
                 offset);
-        state.addOverflow(overflow);
+        state.setOverflow(overflow + state.getOverflow());
```
Yes, although in practice this is very unlikely to fail. You'd need to have a sum that exceeds ~2^190 (in the best case, you'd need to sum 2^64 128-bit numbers to get there). I'll fix it for clarity, though.

martint commented on core/trino-main/src/main/java/io/trino/type/DecimalCasts.java:
```diff
@@ -166,9 +162,9 @@ public static boolean shortDecimalToBoolean(long decimal, long precision, long s
 
     @UsedByGeneratedCode
-    public static boolean longDecimalToBoolean(Slice decimal, long precision, long scale, BigInteger tenToScale)
+    public static boolean longDecimalToBoolean(Int128 decimal, long precision, long scale, BigInteger tenToScale)
```
Yes, but I'll do that separately.

martint commented on core/trino-main/src/main/java/io/trino/type/DecimalOperators.java:
```diff
@@ -106,67 +101,48 @@ public static long addShortShortShort(long a, long b, long aRescale, long bResca
+    public static Int128 addShortShortLong(long a, long b, int rescale, boolean left)
     {
-        return internalAddLongLongLong(unscaledDecimal(a), unscaledDecimal(b), rescale, left);
+        return internalAddLongLongLong(a >> 63, a, b >> 63, b, rescale, left);
```
This is not just really sign bit -- it's computing the leading bits of the number based on its sign (if it's a positive, all 0s. If it's a negative, all 1s)

martint commented on core/trino-main/src/main/java/io/trino/type/DecimalOperators.java:
```diff
@@ -106,67 +101,48 @@ public static long addShortShortShort(long a, long b, long aRescale, long bResca
-            Slice left = unscaledDecimal();
-            Slice right;
-
+            long[] result = new long[2];
```
We need arrays because some of the decoders (ORC, I believe) decode directly into an array of longs. I didn't want to introduce two variants (one with array, one with MutableInt128) quite yet.

martint commented on core/trino-main/src/main/java/io/trino/type/DecimalOperators.java:
```diff
@@ -390,83 +364,99 @@ public static long divideShortShortShort(long dividend, long divisor, int rescal
         try {
-            return divideRoundUp(dividend, rescaleFactor, divisor);
+            Int128 result = divideRoundUp(dividend >> 63, dividend, rescaleFactor, divisor >> 63, divisor, 0);
+            if (Decimals.overflows(result)) {
```
Yes, that's already handled.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128.java:
```diff
@@ -0,0 +1,185 @@
+    private final long high;
+    private final long low;
+
+    public Int128(long high, long low)
```
Not really. Leftover from some earlier version

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1341 @@
+        return ((~a & b) | (~(a ^ b) & (a - b))) >>> 63;
+    }
+
+    public static void subtract(long leftHigh, long leftLow, long rightHigh, long rightLow, long[] result, int offset)
```
No, nothing in the current code would use those variants. We actually don't need the Int128 variant for `add`, so I'll remove it.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1341 @@
+        return high + (low == ALL_BITS_SET_64 ? 1 : 0);
+    }
+
+    private static void incrementUnsafe(long[] value, int offset)
```
It doesn't perform overflow checks -- that's why it's unsafe.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1341 @@
+     * This an unsigned operation. Supplying negative arguments will yield wrong results.
+     * Assumes left array length to be >= 8. However only first 4 int values are multiplied
+     */
+    static void multiply256Destructive(int[] left, Int128 right)
```
This is pre-existent. I have other changes in the pipeline to revamp how we do division, so this will go at that point.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1341 @@
+    }
+
+    // visible for testing
+    public static void shiftLeft(long[] decimal, int shift)
```
It's pre-existent. I'll change it once I get the new division code in.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1341 @@
+        }
+    }
+
+    static void shiftRight(long high, long low, int shift, boolean roundUp, long[] result, int offset)
```
It's pre-existent. I'll change it once I get the new division code in.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1341 @@
+            low = low << shift;
+        }
+        else {
+            high = low << (shift - 64);
```
It's pre-existent. I'll change it once I get the new division code in. However, I believe it matters, as it's part of the division algorithm. Otherwise, we'd need to special case those conditions.

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1341 @@
+
+        dividePositives(dividendHigh, dividendLow, dividendScaleFactor, divisorHigh, divisorLow, divisorScaleFactor, quotient, remainder);
+
+        if (quotientIsNegative) {
+            // negateExact not needed since all positive values can be negated without overflow
+            negate(quotient, 0);
+        }
```
Good catch

martint commented on core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java:
```diff
@@ -0,0 +1,1341 @@
+
+    /**
+     * Divides mutableDividend / mutable divisor
+     * Places quotient in first argument and reminder in first argument
```
It's preexistent, but I'll adjust the comment.

beliefer commented:
@martint Excuse me. I have a question. According to you provided micro benchmark. In some cases, the addition, multiply and division operators have regression. Do the trino fixed these issue? or how to fix them.
